### PR TITLE
Support per-target Prediction

### PR DESCRIPTION
**Pull request by negvet**
Fix issue with map selection/save for detection methods.
Support more smooth overlay of prediction info, if available.

image

### CODE REVIEW

File: openvino_xai/explainer/explanation.py [hunk 1]
```diff
@@ -36,8 +37,9 @@ def __init__(
         self,
         saliency_map: np.ndarray | Dict[int | str, np.ndarray],
         targets: np.ndarray | List[int | str] | int | str,
+        task: Task,
         label_names: List[str] | None = None,
-        metadata: Dict[Task, Any] | None = None,
+        predictions: Dict[Task, Prediction] | None = None,
```
Comment (goodsong81) [inline openvino_xai/explainer/explanation.py]:
```suggestion
        predictions: Dict[int, Prediction] | None = None,
```
If I understand correctly.

**Review by goodsong81 — COMMENTED** [review]
Per-target `Prediction` which is task agnostic <- Looks good to me :)

File: openvino_xai/explainer/explanation.py [hunk 2]
```diff
@@ -57,10 +59,14 @@ def __init__(
             self.layout = Layout.MULTIPLE_MAPS_PER_IMAGE_GRAY
 
         if not explains_all(targets) and not self.layout == Layout.ONE_MAP_PER_IMAGE_GRAY:
-            self._saliency_map = self._select_target_saliency_maps(targets, label_names)
+            if task == Task.DETECTION:
+                self._saliency_map = self._select_target_saliency_maps(targets, None)
+            else:
+                self._saliency_map = self._select_target_saliency_maps(targets, label_names)
```
Comment (GalyaZalesskaya) [inline openvino_xai/explainer/explanation.py]:
```suggestion
            label_names_ = label_names if  task == Task.DETECTION else None
            self._saliency_map = self._select_target_saliency_maps(targets, label_names_)
```

File: openvino_xai/explainer/visualizer.py [hunk 3]
```diff
@@ -162,27 +166,48 @@ def visualize(
...
-                        org=box_label_loc,
-                        fontFace=1,
-                        fontScale=1,
-                        color=(255, 0, 0),
-                        thickness=2,
-                    )
-        return saliency_map_np
+    def _put_classification_info(saliency_map_np: np.ndarray, indices: List[int | str], label_names: List[str] | None, predictions: Dict[int, Prediction]) -> None:
+        for smap, target_index in zip(range(len(saliency_map_np)), indices):
+            corner_location = 3, 17
```
Comment (GalyaZalesskaya) [inline openvino_xai/explainer/visualizer.py]: `corner_location` can be initializid before `for` loop

File: openvino_xai/explainer/visualizer.py [hunk 4]
```diff
@@ -162,27 +166,48 @@ def visualize(
...
+    @staticmethod
+    def _put_detection_info(saliency_map_np: np.ndarray, indices: List[int | str], label_names: List[str] | None, predictions: Dict[int, Prediction]) -> None:
+        for smap, target_index in zip(range(len(saliency_map_np)), indices):
+            saliency_map = saliency_map_np[smap]
+            label_index = predictions[target_index].label
+            score = predictions[target_index].score
+            box = predictions[target_index].bounding_box
+            
+            x1, y1, x2, y2 = box
+            cv2.rectangle(saliency_map, (int(x1), int(y1)), (int(x2), int(y2)), color=(255, 0, 0), thickness=2)
```
Comment (GalyaZalesskaya) [inline openvino_xai/explainer/visualizer.py]:
```suggestion
            x1, y1, x2, y2 = np.array(box, dtype=np.int32)
            cv2.rectangle(saliency_map, (x1, y1), (x2, y2), color=(255, 0, 0), thickness=2)
```

File: openvino_xai/explainer/visualizer.py (hunk 3 above)
Comment (negvet) [reply openvino_xai/explainer/visualizer.py]: Fixed

File: openvino_xai/explainer/visualizer.py (hunk 4 above)
Comment (negvet) [reply openvino_xai/explainer/visualizer.py]: Thx, updated

**Review by goodsong81 — APPROVED** [review]
LGTM :)

**Review by GalyaZalesskaya — APPROVED** [review]

**Pull request merged**